sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,62 +32,109 @@
-        context.state().access(PaneInfoTracker.PANE_INFO_TAG);
-    final Instant windowMaxTimestamp = context.window().maxTimestamp();
-
-    return new ReadableState<PaneInfo>() {
-      @Override
-      public ReadableState<PaneInfo> readLater() {
-        previousPaneFuture.readLater();
-        return this;
-      }
-
-      @Override
-      public PaneInfo read() {
-        PaneInfo previousPane = previousPaneFuture.read();
-        return describePane(key, windowMaxTimestamp, previousPane, isFinal);
-      }
-    };
+    if (hasState(context)) {
+      // We'll need any previous pane so as to derive the next pane's info.
+      final Instant windowMaxTimestamp = context.window().maxTimestamp();
+      final Object key = context.key();
+      final ReadableState<PaneInfo> previousPaneFuture =
+          context.state().access(PaneInfoTracker.PANE_INFO_TAG);
+
+      return new ReadableState<PaneInfo>() {
+        @Override
+        public ReadableState<PaneInfo> readLater() {
+          previousPaneFuture.readLater();
+          return this;
+        }
+
+        @Override
+        public PaneInfo read() {
+          return describePane(key, windowMaxTimestamp, previousPaneFuture.read(), isFinal);
+        }
+      };
+    } else {
+      // For the GlobalWindow calculate the pane without requiring access to any previous pane.
+      return new ReadableState<PaneInfo>() {
+        @Override
+        public ReadableState<PaneInfo> readLater() {
+          // Nothing to read.
+          return this;
+        }
+
+        @Override
+        public PaneInfo read() {
+          // Since this is for the Global window:
+          //  - never first.
+          //  - always early.
+          //  - random nonce for speculative index.
+          //  - always zero for non-speculative index.
+          return PaneInfo.createPane(false, isFinal, isFinal ? Timing.ON_TIME : Timing.EARLY,
+                                     ThreadLocalRandom.current().nextLong(), 0L);
+        }
+      };
+    }
   }
 
+  /**
+   * If required store {@code currentPane} so that any successor pane can build from it.
+   */
   public void storeCurrentPaneInfo(ReduceFn<?, ?, ?, ?>.Context context, PaneInfo currentPane) {
-    context.state().access(PANE_INFO_TAG).write(currentPane);
+    if (hasState(context)) {
+      context.state().access(PANE_INFO_TAG).write(currentPane);
+    }
+    // else: nothing to store for GlabalWindow.
   }
 
-  private <W> PaneInfo describePane(
-      Object key, Instant windowMaxTimestamp, PaneInfo previousPane, boolean isFinal) {
+  /**
+   * Derive the current pane info by combining the previous pane info and state of trigger.
+   */
+  private PaneInfo describePane(
+      Object key, Instant windowMaxTimestamp, @Nullable PaneInfo previousPane, boolean isFinal) {
     boolean isFirst = previousPane == null;
     Timing previousTiming = isFirst ? null : previousPane.getTiming();
     long index = isFirst ? 0 : previousPane.getIndex() + 1;
```

Comment:
Let's switch to an id from index if we can, remove the need for state wholesale, and remove the synchronization point, generally being more friendly to a diversity of implementation strategies.